python/cudf/udf_cpp/groupby/function.cu

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuda/atomic>
+
+#include <cooperative_groups.h>
+
+#include <limits>
+
+template <typename T>
+__device__ void device_sum(cooperative_groups::thread_block const& block,
+                           T const* data,
+                           int64_t size,
+                           T* sum)
+{
+  T local_sum = 0;
+
+#pragma unroll
+  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
+    local_sum += data[idx];
+  }
+
+  cuda::atomic_ref<T, cuda::thread_scope_device> ref{*sum};
+  ref.fetch_add(local_sum, cuda::std::memory_order_relaxed);
+
+  block.sync();
+}
+
+template <typename T>
+__device__ void device_var(cooperative_groups::thread_block const& block,
+                           T const* data,
+                           int64_t size,
+                           double* var)
+{
+  T local_sum      = 0;
+  double local_var = 0;
+
+  __shared__ T block_sum;
+  if (block.thread_rank() == 0) { block_sum = 0; }
+  block.sync();
+
+  device_sum<T>(block, data, size, &block_sum);
+
+  auto const mean = static_cast<double>(block_sum) / static_cast<double>(size);
+
+#pragma unroll
+  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
+    auto temp = static_cast<double>(data[idx]) - mean;
+    temp *= temp;
+    local_var += temp;
+  }
+
+  cuda::atomic_ref<double, cuda::thread_scope_device> ref{*var};
+  ref.fetch_add(local_var, cuda::std::memory_order_relaxed);
+  block.sync();
+
+  *var = *var / static_cast<double>(size - 1);
+  block.sync();
+}
+
+template <typename T>
+__device__ T BlockSum(T const* data, int64_t size)
+{
+  auto block = cooperative_groups::this_thread_block();
+
+  __shared__ T block_sum;
+  if (block.thread_rank() == 0) { block_sum = 0; }
+  block.sync();
+
+  device_sum<T>(block, data, size, &block_sum);
+  return block_sum;
+}
+
+template <typename T>
+__device__ T BlockMean(T const* data, int64_t size)
+{
+  auto block = cooperative_groups::this_thread_block();
+
+  __shared__ T block_sum;
+  if (block.thread_rank() == 0) { block_sum = 0; }
+  block.sync();
+
+  device_sum<T>(block, data, size, &block_sum);
+  return block_sum / static_cast<T>(size);
```

Comment:
So we always turn input data into `double` results for variance and standard deviation, but return the type of the input for `mean` (which is also in general not an integral value). Is this a deliberate choice?

Reply:
FWIW, `mean` is typed to return `float64` in python:

```python
    resolve_mean = _create_reduction_attr(
        "GroupType.mean", retty=types.float64
    )
```

So it's likely numba is inserting a cast into the IR at some point here.

Reply:
Yes the `mean` should be `double`, let me fix this

Reply:
If this should return `double` then shouldn't the return be the following?
```
return static_cast<double>(block_sum) / static_cast<double>(size);
```